parlai/agents/seq2seq/modules.py

```diff
@@ -599,49 +551,81 @@ def __init__(self, attn_type, hidden_size, emb_size, bidirectional=False,
+        bsz, seqlen, hszXnumdir = enc_out.size()
+        numlayersXnumdir = last_hidden.size(1)
+
         if self.attention == 'local':
-            if enc_out.size(1) > self.max_length:
-                offset = enc_out.size(1) - self.max_length
-                enc_out = enc_out.narrow(1, offset, self.max_length)
+            # local attention weights aren't based on encoder states
             h_merged = torch.cat((xes.squeeze(1), last_hidden), 1)
             attn_weights = F.softmax(self.attn(h_merged), dim=1)
-            if attn_weights.size(1) > enc_out.size(1):
-                attn_weights = attn_weights.narrow(1, 0, enc_out.size(1))
+
+            # adjust state sizes to the fixed window size
+            if seqlen > self.max_length:
+                offset = seqlen - self.max_length
+                enc_out = enc_out.narrow(1, offset, self.max_length)
+                seqlen = self.max_length
+            if attn_weights.size(1) > seqlen:
+                attn_weights = attn_weights.narrow(1, 0, seqlen)
         else:
             hid = last_hidden.unsqueeze(1)
             if self.attention == 'concat':
-                hid = hid.expand(last_hidden.size(0),
-                                 enc_out.size(1),
-                                 last_hidden.size(1))
+                # concat hidden state and encoder outputs
+                hid = hid.expand(bsz, seqlen, numlayersXnumdir)
                 h_merged = torch.cat((enc_out, hid), 2)
+                # then do linear combination of them with activation
                 active = F.tanh(self.attn(h_merged))
                 attn_w_premask = self.attn_v(active).squeeze(2)
             elif self.attention == 'dot':
-                if hid.size(2) != enc_out.size(2):
+                # dot product between hidden and encoder outputs
+                if numlayersXnumdir != hszXnumdir:
                     # enc_out has two directions, so double hid
                     hid = torch.cat([hid, hid], 2)
-                attn_w_premask = (
-                    torch.bmm(hid, enc_out.transpose(1, 2)).squeeze(1))
+                enc_t = enc_out.transpose(1, 2)
+                attn_w_premask = torch.bmm(hid, enc_t).squeeze(1)
             elif self.attention == 'general':
+                # before doing dot product, transform hidden state with linear
+                # same as dot if linear is identity
                 hid = self.attn(hid)
-                attn_w_premask = (
-                    torch.bmm(hid, enc_out.transpose(1, 2)).squeeze(1))
-            # calculate activation scores
+                enc_t = enc_out.transpose(1, 2)
+                attn_w_premask = torch.bmm(hid, enc_t).squeeze(1)
+
+            # calculate activation scores, apply mask if needed
             if attn_mask is not None:
                 # remove activation from NULL symbols
+                # TODO: is this the best operation?
                 attn_w_premask -= (1 - attn_mask) * 1e20
```

Comment:
Could do `th.masked_fill_(attn_w_premask, -1e20)`. That's what I've seen in fairseq.

Maybe make that `1e20` as a private constant, like `_ALMOST_INFINITY`? Don't think it matters.

Reply:
`self._YUGE`?